workload: fix some missing Err() checks on sql.Query usages

cucaroach opened the pull request:
When calling Next until it returns false there could be error or it
could be the end of results or an error could have occurred and Err must
be called to see which happened.   In cases where we call Next until it
returns false the query is auto closed and its sufficient to call Err
instead of Close.

This fixes all but one of the cases found by the rowserrcheck linter,
there's one false positive in tpcds that appears to be caused by go
routine usage:

pkg/workload/tpcds/tpcds.go:311:26: rows.Err must be checked

Release justification: fixes hidden errors in test code
Release note: None

Fixes #68164

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69768)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
thanks for this cleanup! i believe there's a linter for this too. can we turn that lint rule on after this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-MibZ-_dAYfR1qN_9bDs:b-aaeces)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach)

___
*[pkg/workload/querylog/querylog.go, line 501 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibViqBDmNoHhhyLG-0:-MibViqBDmNoHhhyLG-1:b-y9rjkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/108c410b342ad6a79a67f19bfd0fea06f6313329/pkg/workload/querylog/querylog.go#L501)):*
> ```Go
> 		return err
> 	}
> 	defer rows.Close()
> ```

i believe we should keep this `defer` since it's possible for the loop to exit early before `rows.Next()` returns false, and it's not an error for `rows` to be double-closed

(second heading here http://go-database-sql.org/errors.html)

also, to fully capture all errors, this could become

```
	defer func() { retErr = errors.CombineErrors(retErr, rows.Close()) }()
```
(along with making the function's returned error a named return value called `retErr`)
___
*[pkg/workload/querylog/querylog.go, line 741 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibWe7wDPZcFmlzvd3V:-MibWe7wDPZcFmlzvd3W:brqqu5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/querylog/querylog.go#L741)):*
> ```Go
> 		// INT2/INT4/INT8 because otherwise they are mapped to the same INT type.
> 		columnTypeByColumnName := make(map[string]string)
> 		rows, err := db.Query(fmt.Sprintf("SELECT column_name, data_type FROM [SHOW COLUMNS FROM %s]", tableName))
> ```

similar to above, can we add a
```
	defer func() { resErr = errors.CombineErrors(resErr, rows.Close()) }()
```
___
*[pkg/workload/querylog/querylog.go, line 778 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibWnYw24iySysTMbDy:-MibWnYw24iySysTMbDz:bm2lru7) ([raw file](https://github.com/cockroachdb/cockroach/blob/108c410b342ad6a79a67f19bfd0fea06f6313329/pkg/workload/querylog/querylog.go#L778)):*
> ```Go
> 		var numCols = 0
> 
> 		defer rows.Close()
> ```

let's make this a 
```
	defer func() { retErr = errors.CombineErrors(retErr, rows.Close()) }()
```
___
*[pkg/workload/querylog/querylog.go, line 821 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibWvz1-q6o7YtTWuva:-MibWvz1-q6o7YtTWuvb:b-m10d1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/querylog/querylog.go#L821)):*
> ```Go
> 		}
> 		log.Infof(ctx, "Sampling %s", tableName)
> 		count, err := db.Query(fmt.Sprintf(`SELECT count(*) FROM %s`, tableName))
> ```

would be better to make this `db.QueryRow` instead so that way we don't need to check `rows.Err()` or call `rows.Close()` at all
___
*[pkg/workload/rand/rand.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibXGfj8nDB4prIGDKP:-MibXGfj8nDB4prIGDKQ:b-m7mcu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/108c410b342ad6a79a67f19bfd0fea06f6313329/pkg/workload/rand/rand.go#L154)):*
> ```Go
> 	var numCols = 0
> 
> 	defer rows.Close()
> ```

let's keep the Close() and make it
```
	defer func() { retErr = errors.CombineErrors(retErr, rows.Close()) }()
```
___
*[pkg/workload/rand/rand.go, line 193 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibXQvS9vj_1SWw1olu:-MibXQvS9vj_1SWw1olv:b-kqcpwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/108c410b342ad6a79a67f19bfd0fea06f6313329/pkg/workload/rand/rand.go#L193)):*
> ```Go
> 			return workload.QueryLoad{}, err
> 		}
> 		defer rows.Close()
> ```

let's make it 
```
	defer func() { retErr = errors.CombineErrors(retErr, rows.Close()) }()
```
___
*[pkg/workload/tpcc/checks.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibXhvh1vT66wC04xuX:-MibXhvh1vT66wC04xuY:bwhsfmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L102)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

after each call to `Query` let's add
```
```
	defer func() { retErr = errors.CombineErrors(districtRows, rows.Close()) }()
```

and so on
```

___
*[pkg/workload/tpcc/checks.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibXvrpAUp1A8vGD-BC:-MibXvrpAUp1A8vGD-BD:byg55rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L217)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

let's add 
```
	defer func() { retErr = errors.CombineErrors(retErr, leftRows.Close()) }()
```
___
*[pkg/workload/tpcc/checks.go, line 229 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibY15s4HT2iVsAUzQA:-MibY15s4HT2iVsAUzQB:b-slfdkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L229)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

and here add
```
```
	defer func() { retErr = errors.CombineErrors(retErr, rightRows.Close()) }()
```
```

___
*[pkg/workload/tpcc/checks.go, line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYJ62CWLS_dkHP24k:-MibYJ62CWLS_dkHP24l:b2jp6ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L274)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

let's add
```
	defer func() { retErr = errors.CombineErrors(retErr, firstQuery.Close()) }()
```
___
*[pkg/workload/tpcc/checks.go, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYSja4NiZDS3V-0q4:-MibYSja4NiZDS3V-0q5:b-yjda1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L276)):*
> ```Go
> 	}
> 	if firstQuery.Next() {
> 		return errors.Errorf("left EXCEPT right returned nonzero results.")
> ```

an alternative would be to use `QueryRow` and assert that the error received is `sql.ErrNoRows` (same for below checks too)
___
*[pkg/workload/tpcc/checks.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYZqg--qJJ3N89cQn:-MibYZqg--qJJ3N89cQo:b-qjd7bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L287)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```


let's add
```
	defer func() { retErr = errors.CombineErrors(retErr, secondQuery.Close()) }()
```
___
*[pkg/workload/tpcc/checks.go, line 313 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYgZw1XLt0i6y6KaV:-MibYgZw1XLt0i6y6KaW:b2jp6ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L313)):*
> ```Go
> 	if err != nil {
> 		return err
> 	}
> ```

let's add
```
	defer func() { retErr = errors.CombineErrors(retErr, firstQuery.Close()) }()
```
___
*[pkg/workload/tpcc/checks.go, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYnOr8SOLTjGNefYV:-MibYnOr8SOLTjGNefYW:beum4p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L320)):*
> ```Go
> 		return err
> 	}
> 	secondQuery, err := txn.Query(`
> ```

we still need the `if err != nil` check for the err returned from txn.Query
___
*[pkg/workload/tpcc/checks.go, line 326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibYtEtDtKfRnFMz46b:-MibYtEtDtKfRnFMz46c:b-3d4166) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/tpcc/checks.go#L326)):*
> ```Go
> (SELECT o_w_id, o_d_id, o_id, o_ol_cnt FROM "order"
>   ORDER BY o_w_id, o_d_id, o_id DESC)`)
> ```

let's add
```
	defer func() { retErr = errors.CombineErrors(retErr, secondQuery.Close()) }()
```


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[pkg/workload/querylog/querylog.go, line 501 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibViqBDmNoHhhyLG-0:-Mibm_Yd4JT7wqAAS6B7:b-s2s74c) ([raw file](https://github.com/cockroachdb/cockroach/blob/108c410b342ad6a79a67f19bfd0fea06f6313329/pkg/workload/querylog/querylog.go#L501)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i believe we should keep this `defer` since it's possible for the loop to exit early before `rows.Next()` returns false, and it's not an error for `rows` to be double-closed

(second heading here http://go-database-sql.org/errors.html)

also, to fully capture all errors, this could become

```
	defer func() { retErr = errors.CombineErrors(retErr, rows.Close()) }()
```
(along with making the function's returned error a named return value called `retErr`)
</blockquote></details>

My first crack at this used to CombineErrors to call Err and Close everywhere, should have stuck with that!   I guess in performance uncritical code doing both is cheap insurance and makes the code less brittle, I'll go back to that model.


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
No, its fires like crazy on false positives.   Unless there was some way to check in a false positive manifest or disable linter in the code.   After my changes this is left in workload:
```
# github.com/cockroachdb/cockroach/pkg/workload/tpcds
pkg/workload/tpcds/tpcds.go:311:26: rows.Err must be checked
```
But there are 119 in pkg/sql/... alone, entirely in _test.go code.   Is there a way to tell the linter to skip tests?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-MiboNxT7JXGsJK2WbZV:bn90zr6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)



<!-- Sent from Reviewable.io -->

cucaroach committed:
workload: fix missing Err() checks on sql.Query usages

When calling Next until it returns false there could be error or it
could be the end of results.  Err must be called to see which happened.
In lots of cases we return errors early and may not ever call Err so
use defer to always call Close.

This fixes all but one of the cases found by the rowserrcheck linter,
there's one false positive in tpcds that appears to be caused by go
routine usage:

pkg/workload/tpcds/tpcds.go:311:26: rows.Err must be checked

Also there are 100+ lint failures in pkg/sql/... test code.  A follow
on PR will investigate how to turn on linting globally.

Release justification: fixes hidden errors in test code
Release note: None
Fixes #68164

cucaroach reviewed (commented):
This is RFAL, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-Mj5fYGp-K3KnHrwLg2n:b-n88pis)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
looks good! i just had minor nits about the style for returning the errors or nils.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-Mj5j_J10U38RPklFa6M:bvwin49)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach)

___
*[pkg/workload/querylog/querylog.go, line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hOsnDGK9AKeSbZpP:-Mj5hOsnDGK9AKeSbZpQ:b-nmmdaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L511)):*
> ```Go
> 	}
> 	retErr = rows.Err()
> 	return retErr
> ```

nit: just use `return rows.Err()` instead
___
*[pkg/workload/querylog/querylog.go, line 814 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hYLh3VDB4A085Bcj:-Mj5hYLh3VDB4A085Bck:bjyseph) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L814)):*
> ```Go
> 		w.state.columnsByTableName[tableName] = cols
> 	}
> 	return retErr
> ```

nit: `return nil` here instead (it makes it more clear that this is the happy case)
___
*[pkg/workload/querylog/querylog.go, line 887 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hf4V1wG3G-voJanV:-Mj5hf4V1wG3G-voJanW:b9jzjjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L887)):*
> ```Go
> 	}
> 	log.Infof(ctx, "Populating samples is complete")
> 	return retErr
> ```

nit: `return nil` here instead
___
*[pkg/workload/rand/rand.go, line 298 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hlsw-rRQkeL6w2AB:-Mj5hlsw-rRQkeL6w2AC:bszvujs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/rand/rand.go#L298)):*
> ```Go
> 		ql.WorkerFns = append(ql.WorkerFns, op.run)
> 	}
> 	return ql, retErr
> ```

nit: `return ql, nil`
___
*[pkg/workload/tpcc/checks.go, line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5iggj2MmHlBwKNvRa:-Mj5iggj2MmHlBwKNvRb:bnsbsev) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/tpcc/checks.go#L259)):*
> ```Go
> 		return errors.Wrap(err, "on `order_line`")
> 	}
> 	return retErr
> ```

nit: `return nil` here
___
*[pkg/workload/tpcc/checks.go, line 298 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5isl-D0vqi9QXtRY2:-Mj5isl-D0vqi9QXtRY3:bd6rm75) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/tpcc/checks.go#L298)):*
> ```Go
> 	}
> 	retErr = secondQuery.Err()
> 	return retErr
> ```

nit: make it `return secondQuery.Err()`
___
*[pkg/workload/tpcc/checks.go, line 342 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5j-Az-tZ87ZDlz2Wi:-Mj5j-Az-tZ87ZDlz2Wj:bd6rm75) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/tpcc/checks.go#L342)):*
> ```Go
> 	}
> 	retErr = secondQuery.Err()
> 	return retErr
> ```

nit: make it `return secondQuery.Err()`


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[pkg/workload/querylog/querylog.go, line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hOsnDGK9AKeSbZpP:-Mj5ktjx4FpmkxWx14lA:b3sdhwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L511)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: just use `return rows.Err()` instead
</blockquote></details>

Hmm, now I'm confused.  Isn't the whole point of using the defer combine errors business to chain the errors into retErr?   In other words why does the defer lambda set retErr if we never use it?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach)

___
*[pkg/workload/querylog/querylog.go, line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hOsnDGK9AKeSbZpP:-Mj5mXow9pn_Y9qzbYh2:b-49r4bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L511)):*
it's a feature of Go -- whatever is returned by the function is automatically assigned to the named return value

e.g. https://play.golang.org/p/7eims9jECuH

there's more info in the language spec: https://golang.org/ref/spec#Return_statements

> Regardless of how they are declared, all the result values are initialized to the zero values for their type upon entry to the function. A "return" statement that specifies results sets the result parameters before any deferred functions are executed. 

we use this pattern elsewhere in our production code as well. e.g. https://github.com/cockroachdb/cockroach/blob/7792137b54f0216514386da5010e4aaed774ebfa/pkg/jobs/registry.go#L939


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[pkg/workload/querylog/querylog.go, line 511 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj5hOsnDGK9AKeSbZpP:-Mj5oodtBAOP6pKIbk8A:bi2fl33) ([raw file](https://github.com/cockroachdb/cockroach/blob/3529efbd26515e2286b61993e95323149edb0d3f/pkg/workload/querylog/querylog.go#L511)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it's a feature of Go -- whatever is returned by the function is automatically assigned to the named return value

e.g. https://play.golang.org/p/7eims9jECuH

there's more info in the language spec: https://golang.org/ref/spec#Return_statements

> Regardless of how they are declared, all the result values are initialized to the zero values for their type upon entry to the function. A "return" statement that specifies results sets the result parameters before any deferred functions are executed. 

we use this pattern elsewhere in our production code as well. e.g. https://github.com/cockroachdb/cockroach/blob/7792137b54f0216514386da5010e4aaed774ebfa/pkg/jobs/registry.go#L939
</blockquote></details>

Sweet, this is exactly what I was missing, thanks!   


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
RFAL, thanks again!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-Mj605jF6Slpg1ocjJYF:b3t8v4w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/workload/querylog/querylog.go, line 821 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-MibWvz1-q6o7YtTWuva:-Mj5r5Eb3Kmg9_PH9tLq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/307dbd1376e7521d225cceb5caecb2bc1a936b57/pkg/workload/querylog/querylog.go#L821)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would be better to make this `db.QueryRow` instead so that way we don't need to check `rows.Err()` or call `rows.Close()` at all
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[pkg/workload/tpcc/checks.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj660lV9f_FiaUI6IPh:-Mj660lW5h0lV5IaZtQh:b-4vp9wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3ec9c1f49000566590ce16b37db0f8791d3e307/pkg/workload/tpcc/checks.go#L305)):*
> ```Go
>   GROUP BY (ol_w_id, ol_d_id, ol_o_id)
>   ORDER BY ol_w_id, ol_d_id, ol_o_id DESC)`)
> 	if err := firstQuery.Scan(); err != gosql.ErrNoRows {
> ```

hm, i think Scan needs the number of arguments to be the same as the number of columns. in this case, there aren't any rows, but it seems like undefined behavior - so better to get the numbers to match up i think. (if that proves to be too cumbersome, then we can go back to Query/rows.Next/rows.Err


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/workload/tpcc/checks.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj660lV9f_FiaUI6IPh:-Mj6WY7k1c6W7IKZcNSQ:b78lnpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3ec9c1f49000566590ce16b37db0f8791d3e307/pkg/workload/tpcc/checks.go#L305)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm, i think Scan needs the number of arguments to be the same as the number of columns. in this case, there aren't any rows, but it seems like undefined behavior - so better to get the numbers to match up i think. (if that proves to be too cumbersome, then we can go back to Query/rows.Next/rows.Err
</blockquote></details>

Well I half tested it and it didn't complain or at least it didn't make any noise but you're right, if there was a row it would return this error:
```
		return fmt.Errorf("sql: expected %d destination arguments in Scan, not %d", len(rs.lastcols), len(dest))
```
so the way I have it written we would silently ignore that error and return the error about having no rows which I think is fine?    Think its too dicey to rely on that behavior?   I could make it more fool proof and avoid dealing with the underlying column types by using the "*RawBytes" argument, should I do that?


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
to get back to the turning on the linter discussion: yeah there is a way to mark test files as excluded.

we do that for TestErrCheck in pkg/testutils/lint/lint_test.go.

this change lgtm though! would you want to do the linting in this PR or separately?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-MjFQPlLC2NTf4Xlii3x:b-oi7l21)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rafiss)

___
*[pkg/workload/tpcc/checks.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj660lV9f_FiaUI6IPh:-MjFQHnN112KFtNbjPC9:baloxks) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3ec9c1f49000566590ce16b37db0f8791d3e307/pkg/workload/tpcc/checks.go#L305)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Well I half tested it and it didn't complain or at least it didn't make any noise but you're right, if there was a row it would return this error:
```
		return fmt.Errorf("sql: expected %d destination arguments in Scan, not %d", len(rs.lastcols), len(dest))
```
so the way I have it written we would silently ignore that error and return the error about having no rows which I think is fine?    Think its too dicey to rely on that behavior?   I could make it more fool proof and avoid dealing with the underlying column types by using the "*RawBytes" argument, should I do that?
</blockquote></details>

i think as-it-is sounds fine. thanks for checking!


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
My inclination is to declare victory here and push this and keep the issue open and do a separate PR for the linting.   Definitely want to put a stake in the ground so we don't regress though.   Thanks for all the help!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-:-MjFTlTB5JfLaf6RLB9P:b-g99vkt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/workload/tpcc/checks.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69768#-Mj660lV9f_FiaUI6IPh:-MjFUJ9J9qa25Ivvq2eE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3ec9c1f49000566590ce16b37db0f8791d3e307/pkg/workload/tpcc/checks.go#L305)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think as-it-is sounds fine. thanks for checking!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

cucaroach commented:
bors r+

cucaroach commented:
bors r+
